internal/message/amqp/publisher.go

```diff
@@ -0,0 +1,172 @@
+package amqp
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/go-logr/logr"
+	"time"
+
+	"github.com/streadway/amqp"
+)
+
+const (
+	reconnectDelay = 5 * time.Second
+	resendDelay = 5 * time.Second
+)
+
+// Publisher represents a connection to a particular queue
+type Publisher struct {
+	name          string
+	logger        logr.Logger
+	connection    *amqp.Connection
+	channel       *amqp.Channel
+	done          chan bool
+	isConnected   bool
+	notifyClose   chan *amqp.Error
+	notifyConfirm chan amqp.Confirmation
+}
+
+// create a new publisher object and try to connect to the queue
+func NewPublisher(uri, queueName string, log logr.Logger) *Publisher {
+	p := Publisher{
+		name:          queueName,
+		logger:        log,
+		done:          make(chan bool),
+	}
+	go p.handleReconnect(uri)
+	return &p
+}
+
+func (p Publisher) handleReconnect(uri string) {
+	for {
+		p.isConnected = false
+		p.logger.Info("Attempting to connect")
+		for !p.connect(uri) {
+			p.logger.Info("Failed to connect. Retrying.")
+			time.Sleep(reconnectDelay)
```

Comment:
not sure if Sleep is the right move here, even if it's for a short period of time (few seconds)